### ISSUE TITLE
(Suspect) Looping too intensively when locked resources selected by custom script are all busy

### ISSUE THREAD

**Issue by jimklimov** [jenkinsci/lockable-resources-plugin#168]
For a few times in the past years severe lags of our Jenkins master (minutes and hours taken to propagate job state changes, to process POSTs in configurations, to find an executor among dozens readily available... all while read requests are snappy and far from all of CPU cores are occupied) happened to correlate with a steady stream of `jenkins.log` entries about lockable resources running our job-defined "groovy expression" scripts to look for a resource to lock. And not finding any because, hey, they are all locked or reserved.

Some of our resources are "assigned" to a dummy user and released from it by a regular health-checking job, so fixing those devices and getting them available as lockable resources quickly makes Jenkins great again.

I suspect that something simple like a longer delay between rounds of checking for all resources (and/or between just individual "checkings" for resources) would free up some locks in the master server to let it also process other tasks.

UPDATE: An alternative gut feeling, based on the assumption that this looping is what blocks our master server from making other decisions and rolling the wheels for jobs unrelated to locked resources at all, while the CPUs are far from clogged, is that maybe the execution of these custom job-defined groovy scripts that identify suitable resources for certain jobs happens without threading it off, so some main master loop blocks for a few seconds (sometimes) until such script completes, does a bit of work, launches another copy of the script, blocks again, ...

**Comment by gmc-devel** [comment]
We have allready observed a similar behaviour; specially, that the UI does not respond (within a few minutes). At least, the selection of resource-canditates by a groovy-script can be very time-consuming and will be repeated every 5 seconds. This will (nearby) block the queue-task and therefore there's to less CPU-time left for other processing ...

To solve this issue, I have built a private variant of this plugin which implements a resource-canditate cache. Now, the mean CPU-usage is bellow 20% instead of nearby 100% (over a long time) if there are a lot of jobs with script-based resource-selection in the queue.

I've provided this suggested fix as PR https://github.com/jenkinsci/lockable-resources-plugin/pull/171 and hope, that it will be taken for an up-coming release ...

**Comment by gmc-devel** [comment]
We have observed a similar bahviour; I.e., the UI does not respond (within a few minutes) if there are many queued jobs which have a scipt-based resource-selection.
At least, I've solved this issue by a private variant of this plugin which implements a canditate-cache. Instead of building a list of resource-canditates for each job all 5 seconds, this cache stores a determined list (individually for each queue-item) for 5 minutes. At least, the CPU-consumption has decreased noticable and the UI works like a charme.

I've provided this suggested fix as PR https://github.com/jenkinsci/lockable-resources-plugin/pull/171 and hope that it will be taken for an up-coming release ...
